experimental/packages/opentelemetry-api-logs/src/types/Logger.ts

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Attributes } from '@opentelemetry/api';
+import { LogRecord } from './LogRecord';
+import { Event } from './Event';
+
+export interface Logger {
```

Comment:
It's not really clear to me how this API is meant to be used. What is the difference between create and emit of a record? Because create has no options it seems like it just creates an empty record and you then have to mutate that record and emit it later? What is the advantage of that over just creating a record with all its data and having it emitted immediately?

Reply:
The intent is to provide a method that creates an instance of LogRecord, which can be held/manipulated until it is ready to be emitted. This follows the pattern of Tracer.startSpan() or Meter.create*(), and also of the existing Logs SDK implementation in Java, which has a LogBuilder pattern. If we have an emit() method that takes an instance of LogRecord, then there must be some way of creating the LogRecord instance through the API.

Reply:
It just doesn't feel very natural to me. I've never seen a logging or event interface that required me to make 2 calls to emit a single record. For tracing and metrics it makes sense because a span has to be ended and a metric can make many measurements.

Reply:
I did include the `emitEvent(name, attributes)` method, which is meant to create and emit at the same time. For events, I could remove `createEvent` and leave only `emitEvent`.

For emitting logs, the issue is that LogRecord has many fields, so the emit method would need many arguments. 

The SDK implementation in Java uses the builder pattern, e.g.
```
logBuilder = logEmitter.logBuilder();
logBuilder.setSeverity(...);
logBuilder.setBody(...);
logBuilder.setAttributes(...);
logBuilder.emit();
```

Alternative options I can think of -

emit method with many arguments
```
emit(timestamp, severityNumber, severityText, body, attributes, flags, traceId, spanId);
```

emit method with options object for optional fields (I think all fields are optional)
```
emit({ severityNumber: 1, body: '', attributes: attributes })
```

create an instance of LogRecord, which has emit() method
```
logRecord = createLogRecord();
logRecord.body = '...';
logRecord.attributes = attributes;
logRecord.emit();
```

Reply:
> emit method with options object for optional fields (I think all fields are optional)
> 
> ```
> emit({ severityNumber: 1, body: '', attributes: attributes })
> ```

This option feels the most natural to me tbh and has the most similarity with our existing APIs. We haven't really used the builder pattern anywhere else. It makes sense for java to use builder pattern because they use it in many places.

Reply:
I think that makes sense. I can leave the LogRecord interface as an argument in the `emit` function. Users of the API will just be passing in an object literal instead of instance of a class. The only downside is that I have `setAttribute` method on the LogRecord interface, which will need to be removed. Attributes will need to be passed in all at once as a nested object.

Reply:
@dyladan @scheler A few more options before I make changes ...

Given these two interfaces for LogRecord and Event
```
interface LogRecord {
   timestamp?: number;
   severityNumber?: number;
   severityText?: string;
   body?: string;
   attributes?: Attributes;
   flags?: number;
   traceId?: string;
   spanId?: string;
}

interface Event {
  name: string;
  domain?: string;
  attributes?: Attributes;
}
```

option 1: two corresponding methods with different names:
```
emitLogRecord(logRecord: LogRecord): void;
emitEvent(event: Event): void;
```

option 2: single overloaded method:
```
log(logRecord: LogRecord): void;
log(event: Event);
```

option 3: emitEvent with event name/domain as separate arguments and optional LogRecord interface argument, which would eliminate the need for the Event interface:
```
emitEvent(eventName: string, eventDomain?: string, logRecord?: LogRecord): void;
```

My preference, I think, would be option 2.

(Note that in option 2 I am calling the method `log` instead of `emit`, since the class is called Logger, not Emitter. I did not do that in option 1 because `logLogRecord` or `logRecord(LogRecord)` seems confusing to me.)

Reply:
I think for low-level logger interface any of these options are fine, but what users really want is typical logging API. Something like
```typescript
interface WrappedLogger {
    debug(body: string, context?:  Omit<LogRecord, "body">): Promise<void>;
    info(body: string, context?: Omit<LogRecord, "body">): Promise<void>;
    warn(body: string, context?: Omit<LogRecord, "body">): Promise<void>;
    error(body: string, context?: Omit<LogRecord, "body">): Promise<void>;
}
```

Especially because we already have `DiagLogger` https://github.com/open-telemetry/opentelemetry-js-api/blob/main/src/diag/types.ts#L27 and I would want to  `diag.setLogger(wrappedLogger);`

So I think it would be nice if there was `WrappedLogger` provided that would use this `Logger` instance under hood. I don't think people would actually want their code base to be full with `logger.log({severityNumber: Severity.INFO, body: 'ehhh'})` it's just really verbose so everyone will write their own wrappers which means it really doesn't matter that much how this `Logger` interface is designed...

Reply:
I prefer option 1 as it is more explicit about what you are trying to do.

> I think for low-level logger interface any of these options are fine, but what users really want is typical logging API. Something like

@davispuh to my understanding this API is not meant to be used by end-users, but by instrumentations which intercept calls from logging libraries which do have the functions you mention. @martinkuba please correct me if I am wrong on that point. Also, it may be a good idea to make that more clear in the TS docs so we don't confuse people.

Reply:
regarding above `interface LogRecord`: Is it really possible to log an empty record? I would assume that at least one element is mandatory.

Reply:
> 
> @davispuh to my understanding this API is not meant to be used by end-users, but by instrumentations which intercept calls from logging libraries which do have the functions you mention. @martinkuba please correct me if I am wrong on that point. Also, it may be a good idea to make that more clear in the TS docs so we don't confuse people.

My point is if I'm already using OpenTelemetry SDK, it has OTLP exporter and it has Logger, why do I need another library? Only a nice wrapper is needed and that's it. It's outside of scope of this PR, but I think SDK should provide it. And if only Instrumentation libraries will use this Logger interface then it's not that big deal how this API is like since it will be used only in couple of places.

Reply:
@dyladan @davispuh Thanks for the feedback. Daniel is correct that this API is intended for instrumentations and log appenders, not for end users. This is described in the corresponding [spec PR](https://github.com/open-telemetry/opentelemetry-specification/pull/2676) and [OTEP](https://github.com/open-telemetry/oteps/blob/main/text/0202-events-and-logs-api.md). I am going to make the change based on option 1.

Reply:
> regarding above interface LogRecord: Is it really possible to log an empty record? I would assume that at least one element is mandatory.

@Flarna All of the fields are optional in the spec.